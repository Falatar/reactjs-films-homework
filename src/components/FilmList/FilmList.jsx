import React, { Component } from 'react';
import PropTypes from 'prop-types';
import style from './FilmList.scss';
import FilmDescription from '../FilmDescription';
import play from '../../static/play.svg';

class FilmList extends Component {
  constructor(props) {
    super(props);
    this.state = {
      detailsMode: false,
      poster: '',
    };
  }

  componentDidMount() {
    const { imgURL } = this.props;
    this.setImg(imgURL);
  }

  switchMode = () => {
    this.setState((state) => ({
      detailsMode: !state.detailsMode,
    }));
  }

  callModal = () => {
    const { callTrailer, setTrailer, id } = this.props;
    setTrailer(id);
    callTrailer();
  }

  setDefaultImg = () => {
    this.setState({ poster: 'https://cdn.pixabay.com/photo/2016/11/30/12/16/question-mark-1872665_1280.jpg' });
  }

  setImg = (imgURL) => {
    if (!imgURL) this.setDefaultImg();
    else this.setState({ poster: `https://image.tmdb.org/t/p/w500${imgURL}` });
  }

  render() {
    const {
      id, name, tagList, rating, overview,
    } = this.props;
    const { poster } = this.state;
    return (
      <div className={style.film__list}>
        <div className={style.poster}>
          <img
            className={style.wall}
            src={poster}
            alt={id}
          />
        </div>
        <div className={style.description}>
          <FilmDescription
            name={name}
            tagList={tagList}
            rating={rating}
            mode
          />
          <p className={style.overview}>{overview}</p>
        </div>
        <div className={style.play__button}>
          <button type="button" className={style.play} onClick={this.callModal}>
            <svg className={style.svg}>
              <use xlinkHref={play} />
            </svg>
          </button>
        </div>
      </div>
    );
  }
}

FilmList.defaultProps = {
  id: 0,
  name: 'Can\'t find property "name"',
  imgURL: '',
  rating: 0,
  tagList: 'Can\'t find property "tagList"',
  overview: 'Can\'t find property "overview"',
  callTrailer: () => 'can\'t find callTrailer function',
  setTrailer: () => 'can\'t find setTrailer function',
};

FilmList.propTypes = {
  id: PropTypes.number,
  name: PropTypes.string,
  imgURL: PropTypes.string,
  rating: PropTypes.number,
  tagList: PropTypes.string,
  overview: PropTypes.string,
  callTrailer: PropTypes.func,
  setTrailer: PropTypes.func,
};

export default FilmList;
