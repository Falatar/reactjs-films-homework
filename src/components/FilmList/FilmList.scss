.film__list {

    height: 10vw;
    width: 95vw;
    background-color: rgba(0, 0, 0, 0.664);
    margin: 1%;
    font-family: "Helvetica Neue", Helvetica, Arial, sans-serif;
    position: relative;
    display: flex;
    flex-direction: row;
    justify-content: center;
    background-size: cover;

    .poster {

        width: 15%;
        height: 100%;
        margin: 0;
        padding: 0;
        border: 0;

        .wall {
            display: block;
            width: 100%;
            height: 100%;
            background-size: cover;
            margin: 0;
            padding: 0;
            border: 0;
            z-index: 100;
            position: relative;
        }
    }

    .description {
        width: 70%;
        height: 100%;
        margin: 0;
        padding: 1%;
        border: 0;
        display: flex;
        flex-direction: column;
        align-items: center;
        justify-content: center;

        .overview {
            width: 90%;
            height: 46%;
            color: white;
            margin: 2% 5%;
            font-size: 1.2em;
            text-overflow: ellipsis;
            resize: none;
            overflow-y: auto;
            scrollbar-color: deepskyblue #201F27;
        }
    }

    .play__button {
        width: 15%;
        height: 100%;
        margin: 0;
        padding: 0;
        border: 0;
        position: relative;
        display: flex;
        flex-direction: column;
        align-items: center;
        justify-content: center;

        .play {
            color: white;
            background-color: deepskyblue;
            height: 1.5em;
            width: 1.5em;
            border: 0;
            border-radius: 100%;
            font-size: 3.5em;
            outline: none;
            padding: 0;

            .svg {
                width: 100%;
                height: 100%;
                fill: white;
                padding-left: 7%;
            }
        }

        .play:active {
            background-image: -webkit-linear-gradient(bottom, rgba(0, 0, 0, 0), rgba(0, 0, 0, 0.4));
            background-image: -moz-linear-gradient(bottom, rgba(0, 0, 0, 0), rgba(0, 0, 0, 0.4));
            background-image: -o-linear-gradient(bottom, rgba(0, 0, 0, 0), rgba(0, 0, 0, 0.4));
            background-image: -ms-linear-gradient(bottom, rgba(0, 0, 0, 0), rgba(0, 0, 0, 0.4));
            background-image: linear-gradient(bottom, rgba(0, 0, 0, 0), rgba(0, 0, 0, 0.4));
        }

        .watch__now {
            color: white;
            font-weight: 500;
            font-size: 1.15em;
        }

        .info {
            //margin-top: 5%;
            margin-bottom: 1%;
            background: transparent;
            color: white;
            font-size: 1.15em;
            border: 1px solid;
            padding: 2% 5%;
            border-radius: 0.5em;
            outline: none;
        }

        .info:active {
            background: rgba(0, 0, 0, 0.9);
        }
    }
}

@media (max-width: 930px) {
    .play {
        font-size: 2em;
    }
}