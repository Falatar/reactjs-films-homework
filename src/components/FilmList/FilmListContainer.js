import { connect } from 'react-redux';
import { updateTrailerPath, openModal } from '../../modules/singleFilmModule/singleFilmModuleActions';
import FilmList from './FilmList';


export const mapDispatchToProps = (dispatch) => ({
  callTrailer: () => dispatch(openModal()),
  setTrailer: (path) => dispatch(updateTrailerPath(path)),
});

export default connect(
  null,
  mapDispatchToProps,
)(FilmList);
