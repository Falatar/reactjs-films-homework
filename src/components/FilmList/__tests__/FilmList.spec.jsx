import React from 'react';
import renderer from 'react-test-renderer';
import FilmList from '../FilmList';

jest.mock('../../FilmDescription', () => 'FilmDescription');

describe('FilmList', () => {
  it('renders correctly', () => {
    const tree = renderer
      .create(
        <FilmList
          id={0}
          name="name"
          imgURL="url"
          rating={0}
          tagList="tagList"
          overview="overview"
          callTrailer={() => { }}
          setTrailer={() => { }}
        />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('default functions works correctly', () => {
    const component = renderer.create(
      <FilmList />,
    );
    const result1 = component.root.props.callTrailer();
    const result2 = component.root.props.setTrailer();
    expect(result1).toEqual('can\'t find callTrailer function');
    expect(result2).toEqual('can\'t find setTrailer function');
  });
});
