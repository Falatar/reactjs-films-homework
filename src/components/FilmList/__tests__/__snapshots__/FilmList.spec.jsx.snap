// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilmList renders correctly 1`] = `
<div
  className="film__list"
>
  <div
    className="poster"
  >
    <img
      alt={0}
      className="wall"
      src="https://image.tmdb.org/t/p/w500url"
    />
  </div>
  <div
    className="description"
  >
    <FilmDescription
      mode={true}
      name="name"
      rating={0}
      tagList="tagList"
    />
    <p
      className="overview"
    >
      overview
    </p>
  </div>
  <div
    className="play__button"
  >
    <button
      className="play"
      onClick={[Function]}
      type="button"
    >
      <svg
        className="svg"
      >
        <use
          xlinkHref="mocked-image"
        />
      </svg>
    </button>
  </div>
</div>
`;
